### PR TITLE
Convert `Unix{Datagram,Stream}::{set_}passcred()` to per-OS traits

### PR DESCRIPTION
**Pull request by jmillikin**
These methods are the pre-stabilized API for obtaining peer credentials from an `AF_UNIX` socket, part of the `unix_socket_ancillary_data` feature.

Their current behavior is to get/set one of the `SO_PASSCRED` (Linux), `LOCAL_CREDS_PERSISTENT` (FreeBSD), or `LOCAL_CREDS` (NetBSD) socket options. On other targets the `{set_}passcred()` methods do not exist.

There are two problems with this approach:

1. Having public methods only exist for certain targets isn't permitted in a stable `std` API.

2. These options have generally similar purposes, but they are non-POSIX and their details can differ in subtle and surprising ways (such as whether they continue to be set after the next call to `recvmsg()`).

Splitting into OS-specific extension traits is the preferred solution to both problems.

### CODE REVIEW

**Comment by jmillikin** [comment]
Additional context: https://github.com/rust-lang/rust/issues/76915#issuecomment-1118979669

This part of the `unix_socket_ancillary_data` feature is not affected by proposed API changes in [RFC 3430](https://github.com/rust-lang/rfcs/pull/3430).

**Comment by joshtriplett** [comment]
r? libs

**Comment by cuviper** [comment]
r? libs-api

File: library/std/src/os/netbsd/net.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
+//! OpenBSD-specific networking functionality.
```
[the remark this answers is not in the document]
Comment (jmillikin) [reply library/std/src/os/netbsd/net.rs]: Done.

**Comment by jmillikin** [comment]
Updated to fix build failure on Android (https://github.com/rust-lang-ci/rust/actions/runs/8109509999/job/22164854739#step:27:2920)

```
  error: methods `passcred` and `set_passcred` are never used
    --> library/std/src/os/net/linux_ext/socket.rs:17:8
     |
  14 | pub trait UnixSocketExt: Sealed {
     |           ------------- methods in this trait
  ...
  17 |     fn passcred(&self) -> io::Result<bool>;
     |        ^^^^^^^^
  ...
  40 |     fn set_passcred(&self, passcred: bool) -> io::Result<()>;
     |        ^^^^^^^^^^^^
     |
     = note: `-D dead-code` implied by `-D warnings`
     = help: to override `-D warnings` add `#[allow(dead_code)]`
```

**Review by dtolnay — APPROVED** [review]
Thanks!

**Comment by jmillikin** [comment]
Is there any way to run the Android portion of CI before trying to merge?

I've tried to fix the Android test failure based on the failed merge's build logs, but I don't have any Android devices capable of running the Rust test suite. Having to iterate on the build via the full bors merge queue is bothersome to other developers, and takes quite a long time.

**Comment by Amanieu** [comment]
You can edit the CI configuration to run the Android builder for PR CI instead of just for bors. You can find the configuration in `src/ci/github-actions/ci.yml`.

r=me once you've checked that the Android CI passes.

**Comment by dtolnay** [comment]
Arm-android CI succeeded in #122324.

@bors r=Amanieu,dtolnay

**Comment by jmillikin** [comment]
@dtolnay Thank you so much for handling the ARM testing! I had poked a bit at the CI config YAML but didn't have time last week to dig in deeper.

Should the same situation happen again, I'll use your commit https://github.com/rust-lang/rust/pull/122324/commits/ed48ec04cd38488a8c24bcf180db74e86c9bf3e2 as a reference for platform-specific CI.

**Pull request merged**